hashes/src/chacha20poly1305/poly1305.rs

```diff
@@ -0,0 +1,245 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Poly1305 one-time message authenticator from RFC7539.
+//!
+//! Heavily inspired by the ["Donna"](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h) implementation in C
+//! and Loup Vaillant's [Poly1305 design article](https://loup-vaillant.fr/tutorials/poly1305-design).
+
+/// 2^26 for the 26-bit limbs.
+const BITMASK: u32 = 0x03ffffff;
+/// Number is encoded in five 26-bit limbs.
+const CARRY: u32 = 26;
+
+/// Poly1305 authenticator takes a 32-byte one-time key and a message and produces a 16-byte tag.
+///
+/// 64-bit constant time multiplication and addition implementation.
+#[derive(Debug)]
+pub struct Poly1305 {
+    /// r part of the secret key.
+    r: [u32; 5],
+    /// s part of the secret key.
+    s: [u32; 4],
+    /// State used to create tag.
+    acc: [u32; 5],
+    /// Leftovers between adds.
+    leftovers: [u8; 16],
+    /// Track relevant leftover bytes.
+    leftovers_len: usize,
+}
```

Comment:
Since this is a hash function it should follow the API of the other hash functions.

Reply:
Is the general idea here to use the `Hash` and `HashEngine` traits?

Reply:
Yes.